ros_bt_py/src/ros_bt_py/ros_nodes/action.py

```diff
@@ -164,106 +195,148 @@ def _do_setup(self):
-            status = self._do_tick_cancel_running_goal()
-            if status != NodeMsg.SUCCEED:
-                return status
+            self._internal_state = ActionStates.REQUEST_GOAL_CANCELLATION
+            return NodeMsg.SUCCEED
+
         return NodeMsg.RUNNING
 
     def _do_tick_cancel_running_goal(self) -> str:
         if self._running_goal_handle is None:
-            self.logfatal("Result handle future is none! Cannot continue!")
+            self.logwarn(
+                "Goal cancellation was requested, but there is no handle to the running goal!"
+            )
+            self._internal_state = ActionStates.FINISHED
             return NodeMsg.BROKEN
+
         self._cancel_goal_future = self._running_goal_handle.cancel_goal_async()
-        return NodeMsg.RUNNING
+        self._internal_state = ActionStates.WAITING_FOR_GOAL_CANCELLATION
+        return NodeMsg.SUCCEED
 
     def _do_tick_wait_for_cancel_complete(self) -> str:
         if self._cancel_goal_future is None:
-            self.logfatal("Cancel goal handle future is none! Cannot continue!")
+            self.logwarn(
+                "Waiting for goal cancellation to complete, but the future is none!"
+            )
+            self._internal_state = ActionStates.FINISHED
             return NodeMsg.BROKEN
+
         if self._cancel_goal_future.done():
+            self.logdebug("Successfully cancelled goal exectution!")
+
             self._cancel_goal_future = None
             self._running_goal_handle = None
             self._running_goal_future = None
-            self._active_goal = None
+
+            self._internal_state = ActionStates.FINISHED
             return NodeMsg.SUCCEED
         if self._cancel_goal_future.cancelled():
+            self.logdebug("Goal cancellation was cancelled!")
+
             self._cancel_goal_future = None
             self._running_goal_handle = None
             self._running_goal_future = None
             self._active_goal = None
-            return NodeMsg.FAIL
+
+            self._internal_state = ActionStates.FINISHED
+            return NodeMsg.FAILURE
         return NodeMsg.RUNNING
 
     def _do_tick_send_new_goal(self) -> str:
-        if self._running_goal_handle is not None or self._new_goal_future is not None:
-            self.logfatal("Running goal handle handle or new goal future is not none! Cannot continue!")
-            return NodeMsg.BROKEN
-        self._new_goal_future = self._ac.send_goal_async(
+        """Tick to request the execution of a new goal on the action server."""
+        self._new_goal_request_future = self._ac.send_goal_async(
             goal=self._input_goal, feedback_callback=self._feedback_cb
         )
-        if self._input_goal is not None:
-            self._active_goal = self._input_goal
+
+        self._active_goal = self._input_goal
+        self._internal_state = ActionStates.WAITING_FOR_GOAL_ACCEPTANCE
+
         return NodeMsg.SUCCEED
```

Comment:
Why do we succeed here? The Action is not done isn't it?

Reply:
As above, for some of the _do_tick methods, the return status does not indicate the entire tick succeeded, but this internal state succeeded. In this case once the goal was successfully submitted this substate succeeds.